src/daft-logical-plan/src/optimization/rules/extract_window_function.rs

```diff
@@ -0,0 +1,488 @@
+use common_treenode::{Transformed, TreeNode};
+use daft_dsl::{expr::window::WindowSpec, resolved_col, Expr, ExprRef};
+use itertools::Itertools;
+
+use crate::{
+    logical_plan::{LogicalPlan, Project},
+    ops::Window,
+    optimization::rules::OptimizerRule,
+};
+
+#[derive(Debug)]
+pub struct ExtractWindowFunction;
+
+impl Default for ExtractWindowFunction {
+    fn default() -> Self {
+        Self
+    }
+}
+
+impl ExtractWindowFunction {
+    pub fn new() -> Self {
+        Self
+    }
+
+    fn is_window_function_expr(expr: &ExprRef) -> bool {
+        expr.exists(|e| matches!(e.as_ref(), Expr::Window(_, _)))
+    }
+
+    fn contains_window_function(project: &Project) -> bool {
+        project.projection.iter().any(Self::is_window_function_expr)
+    }
+
+    fn extract_window_functions(projection: &[ExprRef]) -> Vec<(ExprRef, WindowSpec)> {
+        let mut result = Vec::new();
+
+        for expr in projection {
+            Self::collect_window_functions(expr, &mut result);
+        }
+
+        result
+    }
+
+    fn collect_window_functions(expr: &ExprRef, result: &mut Vec<(ExprRef, WindowSpec)>) {
+        match expr.as_ref() {
+            Expr::Window(_inner_expr, window_spec) => {
+                result.push((expr.clone(), window_spec.clone()));
+            }
+            _ => {
+                for child in expr.children() {
+                    Self::collect_window_functions(&child, result);
+                }
+            }
+        }
+    }
+
+    fn replace_window_functions(
+        expr: &ExprRef,
+        window_col_mappings: &[(ExprRef, String)],
+    ) -> DaftResult<ExprRef> {
+        let transformed = expr.clone().transform(|e| {
+            for (window_expr, col_name) in window_col_mappings {
+                if Arc::ptr_eq(&e, window_expr) {
+                    return Ok(Transformed::yes(resolved_col(col_name.clone())));
+                }
+            }
+            Ok(Transformed::no(e))
+        })?;
+
+        Ok(transformed.data)
+    }
+}
+
+impl OptimizerRule for ExtractWindowFunction {
+    /// Optimizes a logical plan by extracting window function expressions from projections
+    /// and converting them into explicit Window operations.
+    ///
+    /// Algorithm:
+    /// 1. Find all window functions in a projection
+    /// 2. Group them by their window specifications
+    /// 3. Create a separate Window operation for each group
+    /// 4. Connect these Window operations in a chain
+    /// 5. Replace the original window expressions with column references
+    /// 6. Create a final Project operation with the updated expressions
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform_down(|node| match node.as_ref() {
+            LogicalPlan::Project(project) => {
+                if Self::contains_window_function(project) {
+                    let window_funcs = Self::extract_window_functions(&project.projection);
```

Comment:
Small nit: you are traversing the projection twice here but you could do it just once, and if window_funcs is empty, we know that the projection doesn't contain window functions